### DISCUSSION TITLE
Elasticity tensor for 3D cylinder mesh

### DISCUSSION THREAD

**Discussion by MartimSalgado** [idaholab/moose#18679] [Q&A Modules: Solid mechanics] [answered]
Are the principal directions 1,2,3 for C_{ijkl} associated with material coordinates or the global xyz cartesian reference? If we are modeling a 3D cylinder mesh, the elasticity tensor should be defined in terms of cylindrical coordinates and not global xyz.

**Comment by hugary1995** [comment, marked as answer]
So you are asking for a nonhomogeneous anisotropic elasticity tensor. That is doable -- in moose you can define material properties per qp however you like.

**Reply by MartimSalgado** [reply]
@hugary1995 Thanks. I'm saying that this is a relatively common tensor mechanics problem (I'm modelling a fibrous composite material) and there should be an easy way of telling MOOSE that we are defining [C] in terms of a cylindrical coordinate system.

**Reply by hugary1995** [reply]
Yes, it should be pretty straightforward to implement. In your app, create a material that defines the elasticity tensor. You can request information about the coordinate system, e.g. center axis direction, center axis point, etc., from the user by defining input parameters. You can then transform the current point `_q_point[_qp]` into the cylindrical coordinates, and define the elasticity tensor from that.

You can watch the workshop video on the moose homepage for how to create an app and write a custom material.

**Reply by MartimSalgado** [reply]
@hugary1995 that does seem like a straightforward approach. Need to get more into the _q_point transformation.
I've started developing a ComputeElasticityTensor child material ('ComputeFromCylindricalElasticityTensor'). My approach is to rotate Cijkl twice - once for coordinate transformation, another to take into account each transversely isotropic ply's orientation - and then assign it to the elasticity tensor. Here's the thing: I'm not getting accurate results - quantitatively, since C_ij values do not correspond with the material being modeled, and qualitatively (for instance, C_33 varies with (x,y) for a given ply). Do you see anything wrong with my computeQpElasticityTensor?
PS: I haven't programmed this appropriately, yet, I'm just testing to get results. I assume that the layup of plies is made by rotating the radial direction axis and that the z-axis is aligned with the cylinder's axis.

[ComputeFromCylindricalElasticityTensor.zip](https://github.com/idaholab/moose/files/7094877/ComputeFromCylindricalElasticityTensor.zip)

**Reply by MartimSalgado** [reply]
Maybe it would be even better to develop Stress divergence equations kernels for the r-theta-z coordinated system. If I understood correctly, MOOSE only allows to model 2D or 1D problems using polar coordinates, right?

**Reply by MartimSalgado** [reply]
Hey @hugary1995, have you managed to look into this?

**Reply by dschwen** [reply]
> Maybe it would be even better to develop Stress divergence equations kernels for the r-theta-z coordinated system. If I understood correctly, MOOSE only allows to model 2D or 1D problems using polar coordinates, right?

Those already exist

**Reply by MartimSalgado** [reply]
Thanks @dschwen!
I've read the  ADStressDivergenceRZTensors page and it seems that it only works for axisymmetric problems (equations are shown for 2-D, z and r). What I meant by developing the kernel is to obtain the stress divergence for the full 3D problem.
I forgot to mention that I do not have axisymmetric BCs
